contributors/design-proposals/scheduling/gang-scheduling.md

```diff
@@ -0,0 +1,203 @@
+---
+kep-number: 24
+title: Gang Scheduling
+authors:
+  - "@k82cn"
+owning-sig: sig-scheduling, machine-learning WG
+reviewers:
+  - "@bsalamat"
+  - "@vishh"
+approvers:
+  - "@bsalamat"
+  - "@vishh"
+editor: TBD
+creation-date: 2018-07-03
+last-updated: 2018-08-04
+status: provisional
+---
+
+# Gang Scheduling
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Motivation](#motivation)
+* [Function Detail](#function%20detail)
+  * [API Definitation](#api%20definition)
+  * [Lifecycle Policy](#lifecycle%20policy)
+  * [kube-arbitrator](#kube-arbitrator)
+  * [Customized Controller](#customized%20controller)
+* [Feature Interaction](#feature%20interaction)
+  * [Pod RestartPolicy](#pod%20restartpolicy)
+  * [Controller Manager](#controller%20manager)
+  * [Admission Controller](#admission%20controller)
+* [References](#references)
+
+## Motivation
+
+After the discussion at [Gang-scheduling](https://docs.google.com/document/d/1AUwcvTtULNvow5M9e428FnlvINO1uQ7ojRoTGuTp4DA/edit#heading=h.ckn8nv2jj0xv) proposal, we decide to implement Gang-scheduling API in [kube-arbitrator](https://github.com/kubernetes-incubator/kube-arbitrator) by CRDs. kube-arbitrator focuses on "batch" workload in kubernetes, and will share the same [scheduling frameworks](https://github.com/kubernetes/community/pull/2281) when it's ready. This document is used to provide definition of API object and the scheduler behaviour of gang-scheduling.
+
+## Function Detail
+
+### API Definition
+
+The following requirements are identified during the discussion of this feature:
+
+1. Existing workload can use this feature without (or with a few) configuration changes
+2. Pods of a group/gang may have different `PodSpec` (and/or belong to different collections)
+3. Existing controllers which are responsible for managing life cycle of collections work well with this feature
+
+To meet the requirements above, the following **Kind** is introduced in core part under `scheduling.k8s.io/v1alpha1` **Group**/**Version**.
+
+```go
+// PodGroup defines the scheduling requirement of a pod group
+type PodGroup struct {
```

Comment:
How will this definition meet the requirement `2. Pods of a group/gang may have different `PodSpec` (and/or belong to different collections)`? Will this supports scheduler more than one groups together?

Reply:
Regarding multiple/different PodSpec, user can use multiple job or customized job with multiple podSpec; the requirement is to have same GroupName. here's an example ([QueueJob](https://github.com/kubernetes-incubator/kube-arbitrator/blob/master/pkg/apis/extensions/v1alpha1/job.go#L54)) in kube-arbitrator.

And we also support `TaskPriority` within Job to make sure driver will be started before executor, also preemption within Job. There's another topic about how driver/executor find each other if using gang-scheduling + job; will propose something later, or if you have some suggestion.